tests/virt_autotest/xen_guest_irqbalance.pm

```diff
@@ -0,0 +1,192 @@
+# Maintainer: Julie CAO <JCao@suse.com>
+
+use base "virt_feature_test_base";
+use strict;
+use warnings;
+use testapi;
+use version_utils qw(is_sle);
+use set_config_as_glue;
+use virt_autotest::common;
+use virt_autotest::utils qw(is_kvm_host guest_is_sle wait_guest_online print_cmd_output_to_file);
+use virt_utils qw(upload_virt_logs remove_vm restore_downloaded_guests);
+
+sub run_test {
+    my $self = shift;
+
+    return if is_kvm_host;
+    #save original guest configuration file in case of restore in post_fail_hook()
+    my $downloaded_xml_dir = "/tmp/download_vm_xml";
+    foreach my $guest (keys %virt_autotest::common::guests) {
+        unless (script_run("ls $downloaded_xml_dir/$guest.xml") == 0) {
+            assert_script_run "mkdir -p $downloaded_xml_dir";
+            assert_script_run "virsh dumpxml --inactive $guest > $downloaded_xml_dir/$guest.xml";
+        }
+    }
+
+    foreach my $guest (keys %virt_autotest::common::guests) {
+
+        #irqbalance test run on sle12sp3+ guest
+        next if guest_is_sle($guest, '<=12-sp2');
+
+        record_info("Test $guest");
+        prepare_guest_for_irqbalance($guest);
+
+        # with irqbalance.service is enabled, SMP affinity will be distributed among CPUs automatically even if IRQs are bound to specified CPUs in SLE15SP4
+        assert_script_run "ssh root\@$guest \"systemctl stop irqbalance.service\"" if script_run("ssh root\@$guest \"systemctl status irqbalance\"") eq 0;
+
+        my $nproc = script_output("ssh root\@$guest 'nproc'");
+
+        # Bind all NIC IRQs to cpu0 in guest
+        my @nic_irqs_id = split('\n', script_output("ssh root\@$guest \"grep vif /proc/interrupts | cut -d ':' -f1 | sed 's/^ *//'\""));
+        foreach my $irq_id (@nic_irqs_id) {
+            assert_script_run("ssh root\@$guest \"echo 1 > /proc/irq/$irq_id/smp_affinity\"");
+        }
+        my @affinities_with_binding_cpu0 = get_irq_affinities_from_guest($guest, @nic_irqs_id);
+        foreach (@affinities_with_binding_cpu0) {
+            die "Failed to bind NIC IRQs to CPU0" if $_ ne 1;
+        }
+
+        # Start the irqbalance.service again
+        assert_script_run("ssh root\@$guest \"systemctl start irqbalance\"");
+
+        # Check if the SMP affinities are distributed among CPUs
+        my @affinities_with_irqbalance = get_irq_affinities_from_guest($guest, @nic_irqs_id);
+        # for SLE15SP3 and prior, IRQs are distributed on one cpu core ramdomly.
+        # So it is correct if some of them may happen to run on cpu0
+        # but it would not be correct if all IRQs were bound to cpu0.
+        if (guest_is_sle($guest, '<=15-sp3')) {
+            my $multiply = 1;
+            $multiply *= $_ foreach (@affinities_with_irqbalance);
+	    print "julie: \$multiply=$multiply of affinities for $guest after irqbalance.\n";
+            die "SMP affinities were all bound to CPU0 with irqbalance.service enabled" if $multiply == 1;
+        }
```

Comment:
Why multiply 1

Reply:
to avoid all the numbers are 1. The ideal result is that the number of value = 1 should not exceed 1/<cpu_num>*<nic_irq_number>, but it is not absolute. Using multiply = 1 can only ensure that not all are 1

Reply:
I think 'multiply' just a way to help you get result from multiplying elements in @affinities_with_irqbalance. What you really detect is the mutiplying result of all elements of @affinities_with_irqbalance.

Reply:
a little confused. Do you think 'multiply' is ok or do you have some idea?

Reply:
I mean what you really want is the mutiplying result of all elements of @affinities_with_irqbalance, you need variable multiple=1 to do that. Right ?

Reply:
yes, exactly.